pkg/services/ngalert/schedule/schedule.go

```diff
@@ -139,18 +154,164 @@ func (sch *schedule) Unpause() error {
 }
 
 func (sch *schedule) Run(ctx context.Context) error {
-	sch.wg.Add(1)
+	sch.wg.Add(2)
 
 	go func() {
 		if err := sch.ruleEvaluationLoop(ctx); err != nil {
 			sch.log.Error("failure while running the rule evaluation loop", "err", err)
 		}
 	}()
 
+	go func() {
+		if err := sch.adminConfigSync(ctx); err != nil {
+			sch.log.Error("failure while running the admin configuration sync", "err", err)
+		}
+	}()
+
 	sch.wg.Wait()
 	return nil
 }
 
+// SyncAndApplyConfigFromDatabase looks for the admin configuration in the database and adjusts the sender(s) accordingly.
+func (sch *schedule) SyncAndApplyConfigFromDatabase() error {
+	sch.log.Info("start of admin configuration sync")
+	orgIds, err := sch.adminConfigStore.GetOrgsWithAdminConfiguration()
+	if err != nil {
+		return err
+	}
+
+	q := &models.GetOrgAdminConfiguration{OrgIDs: orgIds}
+	if err := sch.adminConfigStore.GetAdminConfigurations(q); err != nil {
+		return err
+	}
+
+	sch.log.Info("found admin configurations", "count", len(q.Result))
+
+	orgsFound := make(map[int64]struct{}, len(q.Result))
+	sch.sendersMtx.Lock()
+	for _, cfg := range q.Result {
+		orgsFound[cfg.OrgID] = struct{}{} // keep track of the which senders we need to keep.
+
+		existing, ok := sch.senders[cfg.OrgID]
+
+		// If the tenant has no Alertmanager(s) configured and no running sender no-op.
+		if !ok && len(cfg.Alertmanagers) == 0 {
+			sch.log.Debug("no external alertmanagers configured", "org", cfg.OrgID)
+			continue
+		}
+
+		// We have a running sender but no Alertmanager(s) configured, shut it down.
+		if ok && len(cfg.Alertmanagers) == 0 {
+			sch.log.Debug("no external alertmanager(s) configured, sender will be stopped", "org", cfg.OrgID)
+			delete(orgsFound, cfg.OrgID)
+			continue
+		}
+
+		// We have a running sender, check if we need to apply a new config.
+		if ok {
+			if sch.sendersCfg[cfg.OrgID] == cfg.AsSHA256() {
```

Comment:
I wonder if there is a more obvious way to decide if we need to apply a new configuration (given that to the calculation of the checksum are included the created and updated columns) eg by taking advantage of the `updated` column or by checking directly the contents of the `cfg.Alertmanagers` which is an array of strings.
In fact, `sendersCfg` could be just `map[int64][]string`

Reply:
I purposely kept `AsSHA256` with the whole struct to make it generic and easy to change. We would only update every 1 minute at most anyways even if there are multiple saves in the database that would update the updated column.

I can simply change:
```
	_, _ = h.Write([]byte(fmt.Sprintf("%v", ac.Alertmanagers)))
```

If you feel strongly about this though.

Reply:
no strong opinions!